docs/Hooks.rst

```diff
@@ -0,0 +1,143 @@
+.. _hooks:
+
+Hooks
+=====
+
+Since EasyBuild v3.5.0, support is available for implementing *hooks* that can be used
+to customise the behaviour of EasyBuild according to site policies if needed,
+without having having to change the EasyBuild framework or the existing easyblocks.
+
+.. contents::
+    :depth: 3
+    :backlinks: none
+
+.. _hooks_what:
+
+What are hooks?
+---------------
+
+*Hooks* are user-defined functions that are called by the EasyBuild framework at specific times during
+the installation procedure that is performed for each software package being installed.
+
+They can be leveraged to alter or augment the installation procedure in a particular way.
+This is usually done to conform with site-specific policies that are difficult to enforce otherwise,
+but it can also be (ab)used to fix specific problems or inject self-implemented enhancements
+(before you flesh them out in a proper contribution, for example).
+
+Both the ``EasyBlock`` instance and the parsed easyconfig file that are being used
+are fully accessible (and modifiable) from within hook implementations, Hence, this mechanism
+provides a lot of flexibility to change the EasyBuild functionality should you require it,
+without having to modify the codebase of EasyBuild itself.
+
+
+.. _hooks_configuration:
+
+Configuring EasyBuild to use your hook implementations
+------------------------------------------------------
+
+To instruct EasyBuild to use your hook implementations,
+you only need to specify the location of the Python module (``*.py``) that implements them.
+
+This is done via the ``--hooks`` configuration option
+(or equivalently via the ``$EASYBUILD_HOOKS`` environment variable, or via ``hooks = ...``
+in an EasyBuild configuration file, see also :ref:`configuring_easybuild`).
+
+For example::
+
+    eb --hooks=$HOME/my_eb_hooks.py ...
+
+or::
+
+    export EASYBUILD_HOOKS=$HOME/my_eb_hooks.py
+    eb ...
+
+
+.. _hooks_available:
+
+Available hooks
+---------------
+
+Currently, two types of hooks are supported:
+
+* ``start_hook`` and ``end_hook``, which are triggered before starting a new software installation, and right after,
```

Comment:
I thought those hooks were called at the beginning and end of the installation "session" so to say. Not per package. I guess I was wrong, but probably that makes sense. In our case, for instance, we would like to refresh the cache at the end of a "session", not every time a new dependency is installed.

Reply:
I can think of use cases for both, so we should add support for an additional set of hooks `start_session_hook` and `end_session_hook`?

That may call for renaming the existing `start_hook` and `end_hook` to something less ambiguous, like `start_easyconfig_hook` and `end_easyconfig_hook`?

I'm open to suggestions for better names...

Reply:
`init_hook` and `finalize_hook`? And `pre_easyconfig_hook`/`post_easyconfig_hook`?